File: leetcode/208.implement-trie-prefix-tree.ts
// @leet start
class Trie {
  // 27th child is to encode a word terminates here
  private readonly children: (Trie | null)[] = new Array(27).fill(null);

  insert(word: string): void {
    const idx = Trie.getChildIndex(word);

    if (!this.children[idx]) {
      this.children[idx] = new Trie();
    }

    if (word.length) {
      this.children[idx].insert(word.slice(1));
    }
  }

  search(word: string): boolean {
    const idx = Trie.getChildIndex(word);

    return (
      this.children[idx] !== null &&
      (!word.length || this.children[idx].search(word.slice(1)))
    );
  }

  startsWith(prefix: string): boolean {
    if (!prefix.length) {
      return true;
    }

    const idx = Trie.getChildIndex(prefix);

    return (
      this.children[idx] !== null &&
      this.children[idx].startsWith(prefix.slice(1))
    );
  }

  private static getChildIndex(word: string): number {
    return word.length ? word.charCodeAt(0) - "a".charCodeAt(0) : 26;
  }
}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */
// @leet end

File: leetcode/208.implement-trie-prefix-tree.py
# @leet start
class Trie:

    def __init__(self):
        self.children: list[Trie | None] = [None] * 26
        self.terminates = False  # does word terminate here?

    @staticmethod
    def _get_char_idx(ch: str) -> int:
        return ord(ch) - ord("a")

    def _search_helper(self, s: str, require_terminate: bool):
        node = self

        for ch in s:
            idx = Trie._get_char_idx(ch)

            if not node.children[idx]:
                return False

            node = node.children[idx]

        return node.terminates or not require_terminate

    def insert(self, word: str) -> None:
        node = self

        for ch in word:
            idx = Trie._get_char_idx(ch)

            if not node.children[idx]:
                node.children[idx] = Trie()

            node = node.children[idx]

        node.terminates = True

    def search(self, word: str) -> bool:
        return self._search_helper(word, True)

    def startsWith(self, prefix: str) -> bool:
        return self._search_helper(prefix, False)


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
# @leet end
